glommio/src/executor/stall.rs

```diff
@@ -0,0 +1,283 @@
+    timer: Arc<sys::timerfd::TimerFd>,
+    handler: Option<JoinHandle<()>>,
+    id: usize,
+    terminated: Arc<AtomicBool>,
+}
+
+impl StallDetector {
+    #[cfg(feature = "stall-detection")]
+    pub(crate) fn new(executor_id: usize) -> std::io::Result<StallDetector> {
+        lazy_static::initialize(&STALL_SIG_HANDLER);
+
+        let timer = Arc::new(
+            sys::timerfd::TimerFd::new(
+                sys::timerfd::ClockId::CLOCK_MONOTONIC,
+                sys::timerfd::TimerFlags::empty(),
+            )
+            .map_err(std::io::Error::from)?,
+        );
+        let tid = unsafe { nix::libc::pthread_self() };
+        let terminated = Arc::new(AtomicBool::new(false));
+        let timer_handler = std::thread::spawn(enclose::enclose! { (terminated, timer) move || {
+            while timer.wait().is_ok() {
+                if terminated.load(Ordering::Acquire) {
+                    return
+                }
+                unsafe { nix::libc::pthread_kill(tid, nix::libc::SIGUSR1) };
+            }
+        }});
+
+        Ok(Self {
+            timer,
+            handler: Some(timer_handler),
+            id: executor_id,
+            terminated,
+        })
+    }
+
+    pub(crate) fn enter_task_queue(
+        &self,
+        queue_name: String,
+        start: Instant,
+        max_expected_runtime: Duration,
+    ) -> nix::Result<StallDetectorGuard<'_>> {
+        // We consider a queue to be stalling the system if it failed to yield in due
+        // time. For a given maximum expected runtime, we allow a margin of error f 10%
+        // (and an absolute minimum of 10ms) after which we record a stacktrace. i.e. a
+        // task queue has should return shortly after `need_preempt()` returns
+        // true or the stall detector triggers. For example::
+        // * If a task queue has a preempt timer of 100ms the the stall detector
+        // triggers if it doesn't yield after running for 110ms.
+        // * If a task queue has a preempt timer of 100ms the the stall detector
+        // triggers if it doesn't yield after running for 15ms.
+
+        let error_margin =
+            Duration::from_millis((max_expected_runtime.as_millis() as f64 * 0.1) as u64)
+                .max(Duration::from_millis(10));
+        StallDetectorGuard::new(
+            self,
+            queue_name,
+            start,
+            max_expected_runtime.saturating_add(error_margin),
+        )
+    }
+
+    fn arm(&self, threshold: Duration) -> nix::Result<()> {
+        self.timer.set(
+            sys::timerfd::Expiration::OneShot(sys::time::TimeSpec::from(threshold)),
+            sys::timerfd::TimerSetTimeFlags::empty(),
+        )
+    }
+
+    fn disarm(&self) -> nix::Result<()> {
+        self.timer.unset()
+    }
+}
+
+impl Drop for StallDetector {
+    fn drop(&mut self) {
+        let timer_handler = self.handler.take().unwrap();
+        self.terminated.store(true, Ordering::Release);
```

Comment:
there should not be the need to set this terminated variable at all. When the timerfd is disarmed, we should just unregister it. If an event does fire, that means it happened close enough to the threshold, so why not log it?

Reply:
If I disarm the timer then the `read` syscall in the background thread will never return. Likewise, I can't close the fd since the kernel does reference counting on its side (and closing fds used across threads is a bad idea anyway.).

I'm not sure I can't get away without it but let me know if I'm missing your point.

Reply:
it's running on a separate thread. So what if it doesn't return?

When you arm it again on the next run, then it might return. And you have to rearm it at every run anyway

Reply:
The `StallDetector` lifecycle is the same as the `LocalExecutor`. This code runs only when you drop a glommio executor, typically at the end of the application. 

I think it's good etiquette to release all the ressources once we are done with them. 🙂 

Just to be clear, in between task queues the timer is disarmed and rearmed just like you describe it.

Reply:
Ok, I see. That's fine, then, but Relaxed atomics should be enough here.
As a rule of thumb, you only need stricter ordering than relaxed when there are two variables that need to be seen in the same order.

Reply:
Fixed in `9287cf8678118e4231bf8daddacc8afacf69da57`.